### ISSUE TITLE
Tokenise based on Camel Case and/or Delimiters

### ISSUE THREAD

**Issue by david-waterworth** [huggingface/tokenizers#466]
I'm looking for a tokenizer which can split an identifier consisting of "words" (or abbreviations) which are delimited either using camel case or using a separator which isn't standard punctuation (i.e. instead of " " the separator is "-", "_" or "." for example). I also want to split between any alpha and numeric characters.

I think CharBPETokenizer will work if I can use a different pre-tokenizer to the WhitespaceSplit() but I don't see one I can use.

Probably the minimum I need to make it work is to be able to split on any boundary between [A-Za-z] and ![A-Za-z]. I think the camel case split is probably not necessary as the BPE usually "learns" that pretty well.

**Comment by Narsil** [comment]
Hi @david-waterworth,

This is a bit specific, are you using `Python` ?
I'll give the more general answer which is to use a `custom` pre_tokenizer. If you are on the rust backend, you can directly implement the `PreTokenizer` trait.

```python
class CamelCasePretok:                                                             
    def get_state(self, c):                                                         
        if c.islower():                                                            
            return "lower"                                                         
        elif c.isupper():                                                          
            return "upper"                                                         
        elif c.isdigit():                                                          
            return "digit"                                                         
        else:                                                                      
            return "rest"                                                          
                                                                                  
    def split(self, n, normalized):                                                
        i = 0                                                                      
        # states = {"any", "lower", "upper", "digit", "rest"}                      
        state = "any"                                                              
        pieces = []                                                                
        for j, c in enumerate(normalized.normalized):                              
            c_state = self.get_state(c)                                            
            if state == "any":                                                     
                state = c_state                                                    
            if state != "rest" and state == c_state:                               
                pass                                                               
            elif state == "upper" and c_state == "lower":                          
                pass                                                               
            else:                                                                  
                pieces.append(normalized[i:j])                                     
                i = j                                                              
            state = c_state                                                        
        pieces.append(normalized[i:])                                              
        return pieces                                                              
                                                                                   
    def pre_tokenize(self, pretok):                                                
        pretok.split(self.split)                                                   
                                                                                     
camel = PreTokenizer.custom(CamelCasePretok())                                     
                                                                                     
assert camel.pre_tokenize_str("HeyThere!?-ThisIsLife") == [                        
    ("Hey", (0, 3)),                                                               
    ("There", (3, 8)),                                                             
    ("!", (8, 9)),                                                                 
    ("?", (9, 10)),                                                                
    ("-", (10, 11)),                                                               
    ("This", (11, 15)),                                                            
    ("Is", (15, 17)),                                                              
    ("Life", (17, 21)),                                                            
]   
```

I'll add this as a test actually as it's a good example for what `custom` pretokenizers are for.
Keep in mind I'm `pieces == [normalized[0:3], normalized[3:8]....]` is really the important bit, we don't return strings here, but a slice of the normalized string, that helps the underlying algorithm figure out offsets.

**Comment by david-waterworth** [comment]
Thanks @Narsil - yes I'm using Python.

**Comment by david-waterworth** [comment]
@Narsil I tried to use this in a `Tokenizer` pipeline but it hangs, i.e. I followed the example from https://github.com/huggingface/tokenizers/issues/243

```
from tokenizers import Tokenizer, trainers
from tokenizers.pre_tokenizers import PreTokenizer
from tokenizers.models import BPE
from tokenizers.normalizers import Lowercase

custom_tokenizer = PreTokenizer.custom(CamelCasePretok()) 

bpe_tokenizer = Tokenizer(BPE())
bpe_tokenizer.normalizer = Lowercase()
bpe_tokenizer.pre_tokenizer = custom_tokenizer

trainer = trainers.BpeTrainer(special_tokens=['<unk>'])
bpe_tokenizer.train(trainer, files)
```

This hangs on the last line, it works if I use one of the supplied tokenizers. I saw elsewhere some comments regarding customer tokenizers not working due to the multithreaded nature of BpeTrainer but I thought perhaps `PreTokenizer.custom()` may have resolved this - it seems not though? Am I doing something wrong or is this usage not supported?

**Comment by Narsil** [comment]
Seems you need to disable parallelism yes, that's not so surprising as GIL intereferes with Rust way of doing parallelisation:

`TOKENIZERS_PARALLELISM=0 python mycmd.py` 

will work. We need to check how to disable that automatically without Deadlock, again :)

**Comment by Narsil** [comment]
I'll close this a open a new PR.

**Comment by wleizny** [comment]
@Narsil the code works for me. But when I try to save the tokenizer to file, it throws an exception: Exception: Custom PreTokenizer cannot be serialized. The same exception occurs when I use customized normalizers. Is it possible to save custom normalizers or pretokenizers to file?

**Comment by david-waterworth** [comment]
@wleizny Since I raised this issue, huggingface have added a Split pre_tokenizers which works with Regex patterns. I've been using the following for CamelCase (strictly speaking it's PascalCase):

    pre_tokenizers.Split(Regex(r"(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])"), "isolated", False)

It's easier as it serialises correctly.
